### DISCUSSION TITLE
Why do contexts do this?

### DISCUSSION THREAD

**Discussion by DaXcess** [solidjs/solid#713] [Q&A] [answered]
So I needed to use contexts for one of my projects and using the [tutorial](https://www.solidjs.com/tutorial/stores_context?solved) I started working on this. However calling useContext in a child component results in `undefined`.

Here's how my code is laid out:
**index.tsx**
```ts
...

render(
  () => (
    <Router>
      <App />
    </Router>
  ),
  document.getElementById("root") as HTMLElement
);
```
**App.tsx**
```ts
...
import { DiscordProvider } from "./lib/discord/provider";
import { HomePage } from "./pages/Home";

const App: Component = () => {
  return (
    <DiscordProvider>
      <HomePage />
    </DiscordProvider>
  );
};

export default App;
```

**provider.tsx**
```ts
export const DiscordContext = createContext<ContextProps>();

export const DiscordProvider: Component = ({ children }) => {
  ...
  return (
    <DiscordContext.Provider
      value={{
        authorize,
        logout,
        state,
        token: () => token,
        userinfo,
      }}
    >
      {children}
    </DiscordContext.Provider>
  );
}
```

and finally
**Home.tsx**
```ts
...
export const HomePage: Component = () => {
  const discord = useContext(DiscordContext);

  createEffect(() => {
    console.log(discord); // test
  });

  return (
    <>
      <h1>?????</h1>
    </>
  );
};
```

If I try to run this code the console will log: `undefined` as the return value of `useContext`. But even stranger is the fact that if I change my `App.tsx` file to something like this:
```ts
...
import { DiscordProvider } from "./lib/discord/provider";
import { HomePage } from "./pages/Home";

const App: Component = () => {
  return (
    <DiscordProvider>
      {() => {
        createEffect(() => {
          console.log(useContext(DiscordContext));
        });

        return <h1>Test</h1>;
      }}
      <HomePage />
    </DiscordProvider>
  );
};

export default App;
```
The new code in this example actually does print out the correct values, whereas the still present `HomePage` component still prints out `undefined`. Does anybody have a clue as to why this is happening because I am lost.

**Comment by ryansolid** [comment, marked as answer]
The problem is probably here:
```js
export const DiscordProvider: Component = ({ children }) => {
```
Destructuring children creates them outside the provider. Instead you should pass props. And access children under the provider.
```js
export const DiscordProvider: Component = (props) => {
  ...
  return (
    <DiscordContext.Provider
      value={{
        authorize,
        logout,
        state,
        token: () => token,
        userinfo,
      }}
    >
      {props.children}
    </DiscordContext.Provider>
  );
}
```

**Reply by DaXcess** [reply]
That indeed does fix the issue. I keep forgetting that destructuring breaks certain aspects of solid.

**Reply by Ximaz** [reply]
Hey there ! I'm a complete newbie in Solid, although I needed context to not pass different props over my components. You helped me solving the issue but, I'd like some information about how deconstructing the props might break solid ? Is it a Javascript/Typescript behaviour ? Is it a reactive library behaviour (might happen in React, Svelt, etc... as well) ? Is it just proper to Solid ? Is it just about the contexts, or could it apply to something else ?

Should I just avoid deconstructing props in general ?
Thanks in advance. :)

**Reply by wegylexy** [reply]
It doesn't break React, which re-renders the heavy virtual DOM every single time when props change. It breaks Solid because `props.children` in Solid JSX is a JavaScript getter function that track where children are gotten, and deconstruction gets these children right away too early, before and outside the beginning of the current component.

**Reply by ryansolid** [reply]
It's a finne-grained reactivity(Signals thing) thing when using getters or proxies. Vue, Solid, Svelte are similar here mechanically. Svelte gets around it by using a compiler to reverse your destructuring. In a sense it is a limitation of JavaScript with these solutions and the only solve is a more invasive compiler than transforms all your code. Some has made solid-undestructure as a similar compiler plugin, but I generally don't recommend it because this behavior is fundamental and fixing it in some places by default would be confusing. Atleast right now the behavior is consistent.
